File: client/src/game/physics.ts
import {
  BARRIER_OFFSET,
  DEFAULT_DIFFICULTY,
  type Difficulty,
  MAX_SPEED_MS,
  nearestCenterline,
  ROAD_HALF_WIDTH,
  type TrackSample,
} from "@racing/shared";
import { carContact, type CarObstacle } from "./car-collision";
import type { CarInput } from "./input";
import { lerpPose, type Pose } from "./pose-interpolation";

interface SurfaceTuning {
  maxSpeed: number;
  engineAccel: number;
  grassMaxSpeed: number;
  grassFriction: number;
}

// Keep these values and the integration order aligned with the trained policy.
// Top speeds also govern the server's lap plausibility validation.
const TUNING: Record<Difficulty, SurfaceTuning> = {
  easy: {
    maxSpeed: MAX_SPEED_MS.easy,
    engineAccel: 38,
    grassMaxSpeed: 24,
    grassFriction: 1.5,
  },
  medium: {
    maxSpeed: MAX_SPEED_MS.medium,
    engineAccel: 65,
    grassMaxSpeed: 9,
    grassFriction: 6,
  },
  hard: {
    maxSpeed: MAX_SPEED_MS.hard,
    engineAccel: 80,
    grassMaxSpeed: 5,
    grassFriction: 10,
  },
};
const BRAKE_DECEL = 38;
const REVERSE_MAX_SPEED = 14;
const COAST_DECEL = 5;
const DRAG = 0.01;
const GRASS_DECEL = 110;
const STEER_RATE = 1.8;
const WALL_DIST = ROAD_HALF_WIDTH + BARRIER_OFFSET - 1.2;
/** Bounciness of car-to-car hits: 0 kills the closing speed, 1 is a perfect bounce. */
const CAR_RESTITUTION = 0.3;

/** Shared by every `advance` without other cars, so a solo car allocates nothing per frame. */
const NO_OBSTACLES: readonly CarObstacle[] = [];

/** Render frames accumulate time; the simulation always consumes fixed steps. */
export const PHYSICS_STEP = 1 / 120;
export const MAX_STEPS_PER_FRAME = 8;
/** Retain ordinary frame hitches, but discard time beyond this backlog. */
export const MAX_ACCUMULATED_TIME = 0.5;

export class CarPhysics {
  x = 0;
  z = 0;
  heading = 0;
  speed = 0;
  onTrack = true;
  centerIndex = 0;

  private touchingWall = false;
  private stepAccumulator = 0;
  private readonly previousPose: Pose = { x: 0, z: 0, heading: 0, speed: 0 };
  private readonly renderPose: Pose = { x: 0, z: 0, heading: 0, speed: 0 };
  private readonly tuning: SurfaceTuning;
  /** A second car that `predict` steps forward, allocated on first use. */
  private scratch: CarPhysics | null = null;

  constructor(
    private readonly difficulty: Difficulty = DEFAULT_DIFFICULTY,
    private readonly samples: TrackSample[],
  ) {
    this.tuning = TUNING[difficulty];
  }

  spawnAtSample(index: number, lateralOffset: number): void {
    const sample = this.samples[index];
    this.x = sample.x + -sample.dirZ * lateralOffset;
    this.z = sample.z + sample.dirX * lateralOffset;
    this.heading = Math.atan2(sample.dirX, sample.dirZ);
    this.speed = 0;
    this.centerIndex = index;
    this.stepAccumulator = 0;
    this.touchingWall = false;
    this.onTrack = Math.abs(lateralOffset) <= ROAD_HALF_WIDTH + 0.6;
    this.rememberPose();
  }

  private rememberPose(): void {
    this.previousPose.x = this.x;
    this.previousPose.z = this.z;
    this.previousPose.heading = this.heading;
    this.previousPose.speed = this.speed;
  }

  /** Seconds of frame time not yet simulated: the physical state trails the latest frame by this. */
  get backlog(): number {
    return this.stepAccumulator;
  }

  /**
   * Render one fixed step behind simulation, smoothly between completed steps.
   * The returned object is reused; network and lap logic must use physical state.
   */
  getRenderPose(): Readonly<Pose> {
    const amount = Math.min(1, this.stepAccumulator / PHYSICS_STEP);
    return lerpPose(this.previousPose, this, amount, this.renderPose);
  }

  /**
   * Pass the raw frame delta so ordinary stalls catch up to the lap clock.
   * `obstacles` are the other players' cars as drawn this frame; Pacers never collide.
   */
  advance(
    elapsed: number,
    input: CarInput,
    obstacles: readonly CarObstacle[] = NO_OBSTACLES,
  ): void {
    if (!Number.isFinite(elapsed) || elapsed < 0) return;
    this.stepAccumulator = Math.min(this.stepAccumulator + elapsed, MAX_ACCUMULATED_TIME);
    for (
      let steps = 0;
      steps < MAX_STEPS_PER_FRAME && this.stepAccumulator >= PHYSICS_STEP;
      steps++
    ) {
      this.update(PHYSICS_STEP, input);
      this.collide(obstacles);
      this.stepAccumulator -= PHYSICS_STEP;
    }
  }

  /**
   * Where this car will be after holding `input` for `seconds`, stepped on the
   * same fixed-step path as `advance` (without other cars) by a scratch copy, so
   * this car is left untouched. A Jev Live Run asks Jev about this pose, since
   * Jev's answer only lands after the round trip (ADR-0009).
   */
  predict(seconds: number, input: CarInput): Pose {
    const copy = (this.scratch ??= new CarPhysics(this.difficulty, this.samples));
    copy.x = this.x;
    copy.z = this.z;
    copy.heading = this.heading;
    copy.speed = this.speed;
    copy.onTrack = this.onTrack;
    copy.centerIndex = this.centerIndex;
    copy.touchingWall = this.touchingWall;
    const steps = Math.max(0, Math.round(seconds / PHYSICS_STEP));
    for (let step = 0; step < steps; step++) copy.update(PHYSICS_STEP, input);
    return { x: copy.x, z: copy.z, heading: copy.heading, speed: copy.speed };
  }

  /** Direct integration is reserved for fixed-dt simulation and training callers. */
  update(dt: number, input: CarInput): void {
    this.rememberPose();
    this.applyPedals(dt, input);
    this.applySurface(dt);
    this.move(dt, input.steer);
    this.resolveBarrier();
  }

  private applyPedals(dt: number, input: CarInput): void {
    if (input.throttle > 0) this.speed += this.tuning.engineAccel * input.throttle * dt;
    if (input.brake > 0) this.speed -= BRAKE_DECEL * input.brake * dt;
    if (input.throttle === 0 && input.brake === 0) {
      this.speed = approachRest(this.speed, COAST_DECEL * dt);
    }
    this.speed -= this.speed * Math.abs(this.speed) * DRAG * dt;
  }

  private applySurface(dt: number): void {
    const nearest = nearestCenterline(this.x, this.z, this.samples);
    this.onTrack = nearest.dist <= ROAD_HALF_WIDTH + 0.6;
    const limit = this.onTrack ? this.tuning.maxSpeed : this.tuning.grassMaxSpeed;
    if (this.speed > limit) this.speed = Math.max(limit, this.speed - GRASS_DECEL * dt);
    if (!this.onTrack && this.speed !== 0) {
      this.speed = approachRest(this.speed, this.tuning.grassFriction * dt);
    }
    if (this.speed < -REVERSE_MAX_SPEED) this.speed = -REVERSE_MAX_SPEED;
  }

  private move(dt: number, steer: number): void {
    const grip = Math.min(Math.abs(this.speed) / 14, 1) / (1 + Math.abs(this.speed) * 0.015);
    this.heading += steer * STEER_RATE * grip * Math.sign(this.speed || 1) * dt;
    this.x += Math.sin(this.heading) * this.speed * dt;
    this.z += Math.cos(this.heading) * this.speed * dt;
  }

  /**
   * Each client only moves its own car (ADR-0008): this car leaves the overlap and
   * takes its half of an equal-mass impulse, and the other player's client
   * resolves their car against this one as it draws it. Only the component along
   * the heading survives, since the model has no sideways velocity; a side-on
   * shove still moves the car through the positional push.
   */
  private collide(obstacles: readonly CarObstacle[]): void {
    let hit = false;
    for (const other of obstacles) {
      const contact = carContact(this, other);
      if (!contact) continue;
      hit = true;
      const { nx, nz, depth } = contact;
      this.x += nx * depth;
      this.z += nz * depth;
      const forwardX = Math.sin(this.heading);
      const forwardZ = Math.cos(this.heading);
      // The other car's speed is reported by its client; never trust it past this Room's limits.
      const otherSpeed = this.limitSpeed(other.speed);
      const closing =
        (forwardX * this.speed - Math.sin(other.heading) * otherSpeed) * nx +
        (forwardZ * this.speed - Math.cos(other.heading) * otherSpeed) * nz;
      if (closing >= 0) continue;
      const impulse = (-(1 + CAR_RESTITUTION) / 2) * closing;
      this.speed += impulse * (forwardX * nx + forwardZ * nz);
    }
    if (!hit) return;
    // A shove must never outrun the Room's top speed, or the lap reads as implausible.
    this.speed = this.limitSpeed(this.speed);
    this.resolveBarrier();
  }

  private limitSpeed(speed: number): number {
    return Math.min(Math.max(speed, -REVERSE_MAX_SPEED), this.tuning.maxSpeed);
  }

  private resolveBarrier(): void {
    const nearest = nearestCenterline(this.x, this.z, this.samples);
    this.centerIndex = nearest.index;
    if (nearest.dist > WALL_DIST) {
      const sample = this.samples[nearest.index];
      const inverseDistance = 1 / nearest.dist;
      this.x = sample.x + (this.x - sample.x) * inverseDistance * WALL_DIST;
      this.z = sample.z + (this.z - sample.z) * inverseDistance * WALL_DIST;
      if (!this.touchingWall) {
        this.speed *= 0.45;
        this.touchingWall = true;
      }
    } else if (nearest.dist < WALL_DIST - 0.5) {
      this.touchingWall = false;
    }
  }
}

function approachRest(speed: number, deceleration: number): number {
  return Math.abs(speed) <= deceleration ? 0 : speed - Math.sign(speed) * deceleration;
}

File: rl/physics.py
"""
Python port of the car physics and track math, mirrored exactly from
shared/src/track.ts (control points, Catmull-Rom sampling, nearest_centerline)
and client/src/game/physics.ts (CarPhysics.update, difficulty tuning, wall clamp).
rl/tests/test_golden.py replays Node-generated trajectories through this port,
so keep the arithmetic order identical to the TypeScript.

Track-taking functions accept a `samples` centerline and default to Sunset Ridge.
"""

import math
from dataclasses import dataclass

ROAD_HALF_WIDTH = 7
BARRIER_OFFSET = 6
TRACK_DIVISIONS = 512

_SUNSET_RIDGE_CONTROL_POINTS = [
    (-40, -210), (40, -213), (110, -205),
    (175, -180), (215, -120),
    (196, -58), (157, -20), (178, 32),
    (225, 85), (235, 150),
    (195, 200), (125, 185),
    (70, 215), (5, 185), (-60, 215), (-130, 205),
    (-185, 155),
    (-150, 95), (-100, 60), (-105, -5),
    (-160, -35), (-205, -80),
    (-195, -150), (-130, -195),
]

# "Serpent's Coil": an apex at every control point (ADR 0003).
_STORMHAVEN_CONTROL_POINTS = [
    (232, 30), (228, -60), (218, -150), (188, -202), (110, -220), (25, -220),
    (-65, -208), (-150, -188), (-198, -150), (-190, -102), (-150, -72),
    (-110, -100), (-70, -73), (-30, -100), (10, -70), (45, -94), (95, -70),
    (135, -20), (140, 40), (110, 80), (60, 95), (0, 80), (-70, 112),
    (-140, 150), (-95, 172), (-40, 178), (90, 175), (185, 140), (225, 90),
]


def _catmull_rom(p0, p1, p2, p3, t):
    t2 = t * t
    t3 = t2 * t
    return 0.5 * (
        2 * p1
        + (-p0 + p2) * t
        + (2 * p0 - 5 * p1 + 4 * p2 - p3) * t2
        + (-p0 + 3 * p1 - 3 * p2 + p3) * t3
    )


def _sample_track(control_points, divisions=TRACK_DIVISIONS):
    n = len(control_points)
    pts = []
    for s in range(divisions):
        u = (s / divisions) * n
        i = int(u)
        t = u - i
        p0 = control_points[(i - 1) % n]
        p1 = control_points[i % n]
        p2 = control_points[(i + 1) % n]
        p3 = control_points[(i + 2) % n]
        x = _catmull_rom(p0[0], p1[0], p2[0], p3[0], t)
        z = _catmull_rom(p0[1], p1[1], p2[1], p3[1], t)
        pts.append((x, z))

    samples = []
    for idx, (x, z) in enumerate(pts):
        nx, nz = pts[(idx + 1) % divisions]
        dx = nx - x
        dz = nz - z
        length = math.hypot(dx, dz) or 1.0
        samples.append({"x": x, "z": z, "dirX": dx / length, "dirZ": dz / length})
    return samples


TRACK_SAMPLES = _sample_track(_SUNSET_RIDGE_CONTROL_POINTS)
STORMHAVEN_SAMPLES = _sample_track(_STORMHAVEN_CONTROL_POINTS)
TRACKS = {"sunset-ridge": TRACK_SAMPLES, "stormhaven": STORMHAVEN_SAMPLES}


def nearest_centerline(x, z, samples=TRACK_SAMPLES):
    best = 0
    best_d2 = float("inf")
    for i, s in enumerate(samples):
        dx = x - s["x"]
        dz = z - s["z"]
        d2 = dx * dx + dz * dz
        if d2 < best_d2:
            best_d2 = d2
            best = i
    return {"index": best, "dist": math.sqrt(best_d2)}


DIFFICULTY_PHYSICS = {
    "easy":   {"max_speed": 52,  "engine_accel": 38, "grass_max_speed": 24, "grass_friction": 1.5},
    "medium": {"max_speed": 90,  "engine_accel": 65, "grass_max_speed": 9,  "grass_friction": 6},
    "hard":   {"max_speed": 110, "engine_accel": 80, "grass_max_speed": 5,  "grass_friction": 10},
}

BRAKE_DECEL = 38
REVERSE_MAX_SPEED = 14
COAST_DECEL = 5
DRAG = 0.01
GRASS_DECEL = 110
STEER_RATE = 1.8
WALL_DIST = ROAD_HALF_WIDTH + BARRIER_OFFSET - 1.2


@dataclass
class PhysicsState:
    x: float
    z: float
    heading: float
    speed: float
    on_track: bool = True
    center_index: int = 0
    touching_wall: bool = False


def spawn_at_sample(index, lateral_offset, samples=TRACK_SAMPLES):
    """Offset along the left-pointing normal of the direction of travel."""
    s = samples[index]
    return PhysicsState(
        x=s["x"] + -s["dirZ"] * lateral_offset,
        z=s["z"] + s["dirX"] * lateral_offset,
        heading=math.atan2(s["dirX"], s["dirZ"]),
        speed=0.0,
        center_index=index,
    )


def step(state, action, dt, difficulty="medium", samples=TRACK_SAMPLES):
    """One fixed step of CarPhysics.update; returns a new state. action: throttle/brake in 0..1, steer in -1..1."""
    tuning = DIFFICULTY_PHYSICS[difficulty]
    throttle = action["throttle"]
    brake = action["brake"]
    steer = action["steer"]

    x = state.x
    z = state.z
    heading = state.heading
    speed = state.speed
    touching_wall = state.touching_wall

    if throttle > 0:
        speed += tuning["engine_accel"] * throttle * dt
    if brake > 0:
        speed -= BRAKE_DECEL * brake * dt
    if throttle == 0 and brake == 0:
        c = COAST_DECEL * dt
        if abs(speed) <= c:
            speed = 0.0
        else:
            speed -= math.copysign(c, speed)
    speed -= speed * abs(speed) * DRAG * dt

    before = nearest_centerline(x, z, samples)
    on_track = before["dist"] <= ROAD_HALF_WIDTH + 0.6
    limit = tuning["max_speed"] if on_track else tuning["grass_max_speed"]
    if speed > limit:
        speed = max(limit, speed - GRASS_DECEL * dt)
    if not on_track and speed != 0:
        f = tuning["grass_friction"] * dt
        if abs(speed) <= f:
            speed = 0.0
        else:
            speed -= math.copysign(f, speed)
    if speed < -REVERSE_MAX_SPEED:
        speed = -REVERSE_MAX_SPEED

    # No grip at standstill, reduced authority at high speed.
    grip = min(abs(speed) / 14, 1.0) / (1.0 + abs(speed) * 0.015)
    # JS: Math.sign(this.speed || 1) treats 0 as positive.
    speed_sign = 1.0 if speed >= 0.0 else -1.0
    heading += steer * STEER_RATE * grip * speed_sign * dt

    x += math.sin(heading) * speed * dt
    z += math.cos(heading) * speed * dt

    # Barrier: clamp to the wall, one-time speed penalty on first contact.
    after = nearest_centerline(x, z, samples)
    center_index = after["index"]
    if after["dist"] > WALL_DIST:
        s = samples[after["index"]]
        inv = 1.0 / after["dist"]
        x = s["x"] + (x - s["x"]) * inv * WALL_DIST
        z = s["z"] + (z - s["z"]) * inv * WALL_DIST
        if not touching_wall:
            speed *= 0.45
            touching_wall = True
    elif after["dist"] < WALL_DIST - 0.5:
        touching_wall = False

    return PhysicsState(x, z, heading, speed, on_track, center_index, touching_wall)
